Fix VisualShader conversion failing with subresources

ColinSORourke opened the pull request:
(Attempts to) Resolve: https://github.com/godotengine/godot/issues/101375

Visual Shaders can have Subresources such as Textures embedded directly in the Visual Shader, and some users find this desirable to better preview the Visual Shader's behavior. However, this breaks the conversion from Visual Shader to Text GDShader because a Text GDshader cannot have SubResources (the appropriate place for these resources would be as a Shader Parameter on a ShaderMaterial).

This PR aligns with @QbieShay's suggestion that 
> I think that converting to shader a visual shader with embedded textures should just fail and print a warning/error.

And resolves the issue by adding a has_node_embeds function to the VisualShader class. Has_node_embeds iterates over the child nodes, and searches through the object properties for embedded resources. The conversion plugin calls this function to catch and throw appropriate errors/warnings.

Note: ConversionPlugin fail states can cause a crash in FileSystem_Dock.cpp - I've mirrored a catch for that from https://github.com/godotengine/godot/pull/109369 to here.

<img width="868" height="118" alt="Screenshot 2025-08-06 at 1 57 49 PM" src="https://github.com/user-attachments/assets/d8a94720-5a02-4609-abc1-e1d5a210c2f0" />

This PR could unblock https://github.com/godotengine/godot/pull/102918, which could decrease file sizes and improve load times. Would require a little more work to have that PRs code handle a failed conversion.

More ideal solutions could be developed in the future such as:
* Disallowing subresource embeds in VisualShaders, and forcing resource references to go through the appropriate ShaderParameters on a Material Resource
* Automatically making embeds 'unique' or creating a Material with the appropriate parameters upon conversion

But I believe this code is still a necessary step to resolve the core error (Conversion deleting subresources) and to track VisualShader subresources for compatibility with future changes.

ColinSORourke committed:
VisualShader Conversion fails with Embeds

Potentially resolves https://github.com/godotengine/godot/issues/101375

VisualShader now has a has_node_embeds function that runs through it's child nodes to find embedded resources via object properties. Conversion plugin uses this function to catch the error.

KoBeWi commented on scene/resources/visual_shader_nodes.cpp:
```diff
@@ -1082,7 +1092,17 @@ String VisualShaderNodeCurveTexture::get_output_port_name(int p_port) const {
 
 void VisualShaderNodeCurveTexture::set_texture(Ref<CurveTexture> p_texture) {
+	if (!texture.is_null()) {
+		if (texture->is_built_in()) {
```
Why this checks? External resources embeds are also broken.

ColinSORourke replied:
The focus of this fix is preventing a 'destructive' conversion. 

The problems are twofold:
A. VisualShader -> GDShader conversion doesn't work with any embeds because a GDShader is a text file without embeds. Any embed would need to be rebound as ShaderParameters on a ShaderMaterial. This is a fundamental consequence of providing this specific conversion. 
B. VisualShader -> GDShader conversion destroys sub resources embedded in the VisualShader because they have no where to go, a user could lose important data.

Fixing A requires a more drastic change, such as removing the VisualShader -> GDShader Conversion, and possibly replacing it with a VisualShader -> ShaderMaterial with embedded GDShader conversion.

In the meantime, fixing B just requires catching embedded sub-resources, and throwing error so the resources can't be deleted. I think 'Embed_count' was an imprecise name, I mean to keep SubResource count. This check in particular catches the case where an embedded sub resource is replaced with an embedded external resource.

I don't think a true embed count is useful - that value is simpler to keep track of, just whether or not each Ref is null. Tracking builtin resources is necessary currently to prevent the destructive conversion, and I think will be useful in future code working with this topic, as it would also need to differentiate BuiltIn resources from Externals.

ColinSORourke commented on editor/shader/visual_shader_editor_plugin.cpp:
```diff
@@ -8437,6 +8437,7 @@ bool VisualShaderConversionPlugin::handles(const Ref<Resource> &p_resource) cons
 Ref<Resource> VisualShaderConversionPlugin::convert(const Ref<Resource> &p_resource) const {
 	Ref<VisualShader> vshader = p_resource;
 	ERR_FAIL_COND_V(vshader.is_null(), Ref<Resource>());
+	ERR_FAIL_COND_V(vshader->has_node_embeds(), Ref<Resource>());
```
Somewhere I got the wrong idea that the "make unique" button in Inspector converted subresources to external resources. Disregard, that's on me

QbieShay commented:
Came back to this. I wonder how we feel about adding a `hint_default("res://blabla.png")` ?

ColinSORourke commented:
@KoBeWi Assuming you are talking about VisualShaderNodeCustom - VisualShaderNodeCustom cannot have Resource Embeds, nothing in the [current API](https://docs.godotengine.org/en/4.4/classes/class_visualshadernodecustom.html#class-visualshadernodecustom-private-method-get-property-name) provides that functionality unless I am misunderstanding. 

@QbieShay 
> Came back to this. I wonder how we feel about adding a hint_default("res://blabla.png") ?

We talked about this in the Rendering Meeting today (Aug.20) (You were there :) ) and determined hint_default is not a workable feature.

Conclusion we reached in the Rendering Meeting for the issue is to throw a warning popup when the conversion plugin is called via Editor, and fall back to throwing an Error (like the current solution) when the conversion is initiated via code.

ColinSORourke commented:
> The problem with this solution is that VisualShader nodes have to count the embeds manually. If someone adds a new node with a resource, but does not mark embed, it will miss the error.
> No, adding a new visual shader node type in the engine. This PR modifies many classes, and in the future there may be more classes that will need similar checks.

Tracking embeds is necessary to catch and communicate this bad conversion behavior. I think the responsibility of tracking embeds is best assigned to individual node classes. Currently only 6 VisualShaderNode classes can have embedded resources, and I would say safe handling of resources on future node classes is important, and low cost.

I have a hard time envisioning an alternative architecture that defers the responsibility of tracking embeds somewhere else? Managing this information from the Visual Shader Editor Plugin seems like it would require more work than having each node track itself, and I don't know if the Visual Shader resource Class has enough info to do any of this tracking.

ColinSORourke commented:
@KoBeWi My bad! Silly mistake, I've pushed a fix.

I changed an inner loop variable name 'i'->'j' to satisfy the code check that inner loop should not use same name as outer loop, but forget to replace in a few relevant locations. Fixed, tested, works.

KoBeWi commented on editor/shader/visual_shader_editor_plugin.cpp:
```diff
@@ -8446,6 +8446,11 @@ bool VisualShaderConversionPlugin::handles(const Ref<Resource> &p_resource) cons
+	int embed = vshader->has_node_embeds();
+	ERR_FAIL_COND_V_MSG(embed == 2, Ref<Resource>(), "Cannot convert VisualShader to GDShader because VisualShader has embedded subresources.");
+	if (embed == 1) {
+		WARN_PRINT("Visual Shader conversion cannot convert embedded resources. Resource references from Nodes will have to be rebound as ShaderParameters on a Material");
```
"embedded" is not correct here. They are not embedded, they are external dependencies of the shader.

btw are the shader parameters automatically created?

ColinSORourke replied:
Updated the language on the error.

As far as I can tell, yes the Shader Parameters are automatically created. A Texture2D Node (which could have an embedded texture) is converted to a Uniform in the GD Shader Code, which will be recognized as a Shader Parameter.

The VisualShader in the MRP has 2 Texture2D Nodes, they convert to code 
```
uniform sampler2D tex_frg_44;
uniform sampler2D tex_frg_43;
```
And when the GDShader is attached to a ShaderMaterial, the uniforms show up as a parameter again

ColinSORourke commented:
@KoBeWi Updating warning language, and changed to throw a Toaster Message (when it can - I think there are cases where a conversion plugin can be invoked outside of editor)

> Also converting the shader from the inspector results in clearing the property if there are sub-resources.
> This could be avoided by returning the same shader instead of null.

Made this update too, although I think it'd be preferable if the Inspector Dock handled a null return and recognized the value as unchanged.  Looking for the file that manages that ;)

KoBeWi commented:
Looks like easy fix:
https://github.com/godotengine/godot/blob/adb2ec043ed61ea8451f2593171d7163303b7ba5/editor/inspector/editor_resource_picker.cpp#L487

ColinSORourke commented:
Added a catch in Editor_Resource_Picker conversion to handle being given null by a failed conversion. Returned the Conversion plugin to always return null in a failed conversion.

Also added `emit_signal(SNAME("file_removed")` to the FileSystemDock conversion action - without this, if a VisualShader was converted to Shader, the ShaderEditor would keep the VisualShader version open as a hanging [unsaved]

KoBeWi commented on editor/inspector/editor_resource_picker.cpp:
```diff
@@ -484,8 +484,11 @@ void EditorResourcePicker::_edit_menu_cbk(int p_which) {
-				edited_resource = conversions[to_type]->convert(edited_resource);
-				_resource_changed();
+				Ref<Resource> converted_resource = conversions[to_type]->convert(edited_resource);
+				if (!converted_resource.is_null()) {
```
```suggestion
				if (converted_resource.is_valid()) {
```
